crates/sparrow-session/src/session.rs

```diff
@@ -345,31 +362,49 @@ impl Session {
         }
     }
 
+    /// Execute the query.
+    ///
+    /// It is unnfortunate this requires `&mut self` instead of `&self`. It relates to the
+    /// fact that the decorations may require mutating the DFG, which in turn requires
+    /// mutability. In practice, the decorations shouldn't mutate the DFG and/or that
+    /// shouldn't require mutating the session.
```

Comment:
What are the other options? Could we refactor it such that the decorations create their own DFG, then we connect the original + decoration in different step?

Reply:
Not really, since the decorations take the original as their argument. The best option would be if we had logical plans -- then this isn't mutating any DFG, it's just creating a new logical plan node applying the decoration. Then, even if we create a DFG it would be part of compiling the logical plan. Not something I think we should do here.